### PR TITLE
User Model Cleanup & Refactoring

### PR DESCRIPTION
**Pull request by nanafox**
This pull request includes several changes to the Centa User Management Service project, focusing on enhancing the user and role management functionality, improving code readability, and updating documentation. The most important changes include the addition of new models and migrations, updates to the admin configuration, and various improvements to the `CustomUserManager` class.

### Enhancements to User and Role Management:

* **New Models and Migrations:**
  - Added models for `Organisation`, `Permission`, `Role`, `RolePermission`, `User`, and `UserProfile` in `src/user_service/models.py`.
  - Created initial migration file `0001_initial.py` to define the database schema for the new models.

* **Admin Configuration:**
  - Updated `src/user_service/admin.py` to register the new models with the Django admin site.

### Code Readability Improvements:

* **CustomUserManager Class:**
  - Improved the `create_user`, `create_superuser`, `deactivate_user`, `activate_user`, and `update_user` methods to use `ValidationError` instead of `ValueError` and added helper methods for better code organization. [[1]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL25-R101) [[2]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL94-R147)

* **General Code Cleanup:**
  - Removed unnecessary comments and improved docstrings for better clarity in `src/user_service/models.py`. [[1]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL1-R10) [[2]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL122-R292)

### Documentation Update:

* **README.md:**
  - Added instructions for creating a secret key and setting up the `.env` file for local development.

These changes collectively improve the functionality, maintainability, and documentation of the Centa User Management Service project.

### CODE REVIEW

**Comment by thenoblet** [comment]
If a user lookup fails a UserDoesNotExist exception is to be raised. Validation errors are mostly for form validations, model field validations and custom validation logic.

**Comment by thenoblet** [comment]
Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.

What do you think??

File: src/user_service/models.py
Comment (thenoblet) [inline src/user_service/models.py:1]: If a user lookup fails a UserDoesNotExist exception is to be raised. Validation errors are mostly for form validations, model field validations and for custom validation logic

**Review by thenoblet — APPROVED** [review]
Looks good... well done, Mentor!!!

File: src/user_service/models.py
Comment (nanafox) [reply src/user_service/models.py:1]: Before this update, your initial code was raising a `ValueError`. I will update it to raise the appropriate exception now.

**Comment by nanafox** [comment]
> Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.
> 
> What do you think??

We could do that but instead of `ValueError`, it will be an `AttributeError` since we are setting them as attributes of the that user instance.

If this is the case, we will also need to have the same checks in the `create_user` method as well since it also accepts multiple extra fields.

**Comment by nanafox** [comment]
> > Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.
> > What do you think??
> 
> We could do that but instead of `ValueError`, it will be an `AttributeError` since we are setting them as attributes of the that user instance.
> 
> If this is the case, we will also need to have the same checks in the `create_user` method as well since it also accepts multiple extra fields.

Here's an example of how it would go. This example uses the `create_user` method, but it will work with the `update_user` method as well. This way, errors are silent when an invalid attribute is passed to either of these methods.

image

**Comment by thenoblet** [comment]
> > > Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.
> > > What do you think??
> > 
> > 
> > We could do that but instead of `ValueError`, it will be an `AttributeError` since we are setting them as attributes of the that user instance.
> > If this is the case, we will also need to have the same checks in the `create_user` method as well since it also accepts multiple extra fields.
> 
> Here's an example of how it would go. This example uses the `create_user` method, but it will work with the `update_user` method as well. This way, errors are silent when an invalid attribute is passed to either of these methods.
> 
> image

Looks good!

**Pull request merged**